### PR TITLE
Add docs on SQL cursors (DECLARE, FETCH, CLOSE)

### PR DESCRIPTION
**Pull request by rmloveland**
Fixes:

- DOC-2699
- DOC-2700
- DOC-6185

NB. Ported the changes to v22.1, v22.2, v23.1.

Also added a link from the original v22.1 release note message about cursors to the v22.1 version of these docs

### CODE REVIEW

**Comment by rmloveland** [comment]
@dikshant added you as a reviewer but also no worries if you don't have time etc, hunting for an engineering review as well (per [this comment](https://cockroachlabs.atlassian.net/browse/DOC-6185?focusedCommentId=162351) you prolly saw)

File: _includes/v23.1/known-limitations/sql-cursors.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+CockroachDB implements SQL {% if page.name == "known-limitations.md" %} [cursor](cursors.html) {% else %} cursor {% endif %} support with the following limitations:
+
+- `DECLARE` only supports forward cursors.
+- `FETCH` supports forward, relative, and absolute variants, but only for forward cursors.
+- `BINARY CURSOR`, which returns data in the Postgres binary format, is not supported.
+- `MOVE`, which allows advancing the cursor without returning any rows, is not supported.
+- `WITH HOLD`, which allows keeping a cursor open for longer than a transaction by writing its results into a buffer, is accepted as valid syntax within a single transaction but is not supported. It acts as a no-op and does not actually perform the function of `WITH HOLD`, which is to make the cursor live outside its parent transaction.
```
Comment (dikshant) [inline _includes/v23.1/known-limitations/sql-cursors.md:7]: I think important to stress here that if you are using WITH HOLD you will be forced to close that cursor within the transaction it was created in.

File: v23.1/cursors.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+Cursors differ from [keyset pagination](pagination.html) and [`LIMIT`/`OFFSET`](limit-offset.html) in that:
+
+- Each cursor is a stateful SQL object that is referred to by a unique name.
+- Each cursor requires holding open its own dedicated (read-only) [transaction](transactions.html).
+- Each cursor operates on a snapshot of the database at the moment that cursor is opened.
+
+## Synopsis
+
+Cursors are declared using the following keywords:
```
Comment (dikshant) [inline v23.1/cursors.md:17]: Isn't just DECLARE used to declare a cusor? FETCH and CLOSE are used to interact with it not declare it.

File: _includes/v23.1/known-limitations/sql-cursors.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    - This syntax is not accepted, and will result in an error:
+        {% include_cached copy-clipboard.html %}
+        ~~~ sql
+        BEGIN;
+        DECLARE test_cur CURSOR WITH HOLD FOR SELECT * FROM foo ORDER BY bar;
+        COMMIT; -- This will fail with an error because CLOSE test_cur was not called inside the transaction.
+        ~~~
+- Scrollable cursor (also known as reverse `FETCH`) is not supported.
+- [`SELECT ... FOR UPDATE`](select-for-update.html) is not supported.
+- Respect for [`SAVEPOINT`s](savepoint.html) is not supported. Cursor definitions do not disappear properly if rolled back to a `SAVEPOINT` from before they were created.
```
Comment (dikshant) [inline _includes/v23.1/known-limitations/sql-cursors.md:25]: There are a few other things not supported which might be worth adding here:
https://github.com/cockroachdb/cockroach/issues/41412#issuecomment-1053649403

**Review by dikshant — COMMENTED** [review]
Added a few suggestions.

File: _includes/v23.1/known-limitations/sql-cursors.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        ~~~
+    - This syntax is not accepted, and will result in an error:
+        {% include_cached copy-clipboard.html %}
+        ~~~ sql
+        BEGIN;
+        DECLARE test_cur CURSOR WITH HOLD FOR SELECT * FROM foo ORDER BY bar;
+        COMMIT; -- This will fail with an error because CLOSE test_cur was not called inside the transaction.
+        ~~~
+- Scrollable cursor (also known as reverse `FETCH`) is not supported.
+- [`SELECT ... FOR UPDATE`](select-for-update.html) is not supported.
```
Comment (rafiss) [inline _includes/v23.1/known-limitations/sql-cursors.md:24]: could you clarify: "using SELECT FOR UPDATE with a cursor is not supported"

File: v23.1/cursors.md (hunk 2 above)
Comment (rafiss) [reply v23.1/cursors.md:17]: we can change to "Cursors are declared and used with the following keywords" or something

**Review by rafiss — APPROVED** [review]
just some small nits from my end. thanks!

File: _includes/v23.1/known-limitations/sql-cursors.md (hunk 1 above)
Comment (rmloveland) [reply _includes/v23.1/known-limitations/sql-cursors.md]: Thanks, I added this info so it's super clear (hopefully)!

File: v23.1/cursors.md (hunk 2 above)
Comment (rmloveland) [reply v23.1/cursors.md]: Update to the language suggested by Rafi!

File: _includes/v23.1/known-limitations/sql-cursors.md (hunk 3 above)
Comment (rmloveland) [reply _includes/v23.1/known-limitations/sql-cursors.md]: Thank you - I have updated the descriptions of the reverse scrolling stuff a bit, and added the GitHub links

File: v23.1/cursors.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+## Synopsis
+
+Cursors are declared and used with the following keywords:
+
+- [`DECLARE`](sql-grammar.html#declare_cursor_stmt)
+- [`FETCH`](sql-grammar.html#fetch_cursor_stmt)
+- [`CLOSE`](sql-grammar.html#close_cursor_stmt)
+
+<div>
```
Comment (ianjevans) [inline v23.1/cursors.md:23]: I think the bogus Vale failures are caused by the `remote_include` being surrounded by the `<div>` tag.

File: _includes/v23.1/known-limitations/sql-cursors.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+CockroachDB implements SQL {% if page.name == "known-limitations.md" %} [cursor](cursors.html) {% else %} cursor {% endif %} support with the following limitations:
+
+- `DECLARE` only supports forward cursors. Reverse cursors created with `DECLARE SCROLL` are not supported. [cockroachdb/cockroach#77102](https://github.com/cockroachdb/cockroach/issues/77102)
```
Comment (ianjevans) [inline _includes/v23.1/known-limitations/sql-cursors.md:3]: These Vale failures, man. Annoying. If you want to have Vale ignore the link text to CockroachDB GitHub issues, add the following to `vale.ini`:

~~~
TokenIgnores = (\[cockroachdb/cockroach.*?\])
~~~

**Review by ianjevans — APPROVED** [review]
LGTM, just left a couple suggestions on getting Vale to calm the hell down.

File: v23.1/cursors.md (hunk 5 above)
Comment (rmloveland) [reply v23.1/cursors.md:23]: Interesting, I haven't seen that before.  AFAICT we always wrap the SVG diagrams in the `div` tags everywhere else.  Not sure why it's mad here.  Oh well hope the next one passes? :-)

File: _includes/v23.1/known-limitations/sql-cursors.md (hunk 6 above)
Comment (rmloveland) [reply _includes/v23.1/known-limitations/sql-cursors.md:3]: Thanks for this info, added that line to the `.vale.ini` file

File: .vale.ini [hunk 7]
```diff
@@ -30,3 +30,5 @@ BlockIgnores = (?s)(<(?:div|section)[^>]*markdown="1"[^>]*>.*?</(?:div|section)>
 #
 # Rule #1: Ignore all Liquid tags.
 TokenIgnores = (\{%.*?%\})|(\{\{.*?\}\})
+# Rule #2: Ignore links of the form [cockroachdb/cockroach#123](...)
+TokenIgnores = (\[cockroachdb/cockroach.*?\])
```
Comment (nickvigilante) [inline .vale.ini:34]: @rmloveland What errors were you getting that prompted the creation of this?

File: _includes/v23.1/known-limitations/sql-cursors.md (hunk 6 above)
Comment (nickvigilante) [reply _includes/v23.1/known-limitations/sql-cursors.md:3]: I'm gonna investigate this locally. Vale shouldn't be this unnecessarily chatty.

File: .vale.ini (hunk 7 above)
Comment (nickvigilante) [reply .vale.ini:34]: NVM, I found it now.

File: _includes/v23.1/known-limitations/sql-cursors.md (hunk 6 above)
Comment (nickvigilante) [reply _includes/v23.1/known-limitations/sql-cursors.md:3]: Yeah, that's a good fix.

**Pull request merged**